cmd/influxd/inspect/report_tsi/report_tsi.go

```diff
@@ -0,0 +1,481 @@
+		}
+		r.lowCardinality = nil
+	}
+	r.set.Merge(other)
+}
+
+type results []*result
+
+func (a results) Len() int           { return len(a) }
+func (a results) Less(i, j int) bool { return a[i].count < a[j].count }
+func (a results) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func (report *reportTSI) printSummaryByMeasurement() error {
+	// Get global set of measurement names across shards.
+	idxs := &tsdb.IndexSet{SeriesFile: report.sfile}
+	for _, idx := range report.shardIdxs {
+		idxs.Indexes = append(idxs.Indexes, idx)
+	}
+
+	mitr, err := idxs.MeasurementIterator()
+	if err != nil {
+		return err
+	} else if mitr == nil {
+		return errors.New("got nil measurement iterator for index set")
+	}
+	defer mitr.Close()
+
+	var name []byte
+	var totalCardinality int64
+	measurements := results{}
+	for name, err = mitr.Next(); err == nil && name != nil; name, err = mitr.Next() {
+		res := &result{name: name}
+		for _, shardCards := range report.cardinalities {
+			other, ok := shardCards[string(name)]
+			if !ok {
+				continue // this shard doesn't have anything for this measurement.
+			}
+
+			if other.short != nil && other.set != nil {
+				panic("cardinality stored incorrectly")
+			}
+
+			if other.short != nil { // low cardinality case
+				res.addShort(other.short)
+			} else if other.set != nil { // High cardinality case
+				res.merge(other.set)
+			}
+
+			// Shard does not have any series for this measurement.
+		}
+
+		// Determine final cardinality and allow intermediate structures to be
+		// GCd.
+		if res.lowCardinality != nil {
+			res.count = int64(len(res.lowCardinality))
+		} else {
+			res.count = int64(res.set.Cardinality())
+		}
+		totalCardinality += res.count
+		res.set = nil
+		res.lowCardinality = nil
+		measurements = append(measurements, res)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// sort measurements by cardinality.
+	sort.Sort(sort.Reverse(measurements))
+
+	if report.topN > 0 {
+		// There may not be "topN" measurement cardinality to sub-slice.
+		n := int(math.Min(float64(report.topN), float64(len(measurements))))
+		measurements = measurements[:n]
+	}
+
+	tw := tabwriter.NewWriter(os.Stdout, 8, 8, 1, '\t', tabwriter.AlignRight)
+
+	fmt.Fprintf(tw, "Summary\nDatabase Path: %s\nCardinality (exact): %d\n\n", report.enginePath, totalCardinality)
+	fmt.Fprint(tw, "Measurement\t\tCardinality (exact)\n\n")
+	for _, res := range measurements {
+		fmt.Fprintf(tw, "%q\t%d\t\n", res.name, res.count)
+	}
+
+	if err := tw.Flush(); err != nil {
+		return err
+	}
+	fmt.Fprint(os.Stdout, "\n\n")
+	return nil
+}
+
+func (report *reportTSI) printShardByMeasurement(id uint64) error {
+	allMap, ok := report.cardinalities[id]
+	if !ok {
+		return nil
+	}
+
+	var totalCardinality int64
+	all := make(cardinalities, 0, len(allMap))
+	for _, card := range allMap {
+		n := card.cardinality()
+		if n == 0 {
+			continue
+		}
+
+		totalCardinality += n
+		all = append(all, card)
+	}
+
+	sort.Sort(sort.Reverse(all))
```

Comment:
From here down looks very similar to the above, could it be factored out into a function?

Reply:
So I tried factoring this out into something like this:

```
func (report *reportTSI) printSection(info interface{}, path string, totalCardinality int64)error{
	switch info.(type) {
	case results:
		// print summary section
	case cardinalities:
		// print shard section
	}
	return nil
}
```
It seemed like an overly complex fix, and still had quite a bit of code that needed to be repeated. This is because it would need to be able to handle two different types of info to be printed (results struct and cardinalities struct), both of which have different variables and methods associated with them. In the end, I kept this the same as iteration 1. Maybe there's a simpler way to factor it out, so open to suggestions if you think it's important.

Reply:
It's ok to leave as-is :+1: given the choice between living with some duplication vs. complicating the code to avoid it, Go style tends to prefer the duplication (but it's good to sanity-check)